### PR TITLE
(MODULES-3945) Add Non Windows PowerShell platform support

### PR DESCRIPTION
**Pull request by glennsarti**
Previously, the PowerShell module would only function on Windows hosts.  This
commit adds support for running PowerShell on non-Windows hosts with the
following:
- Use the older PowerShell module style functionality with a temporary script
  file
- Use default file locations for PowerShell at /usr/bin and /usr/local/bin
- Guard the use of cmd.exe on Windows only platforms
- Update the spec tests to use Unix fixtures for testing e.g. uname
- Remove the use of -ExecutionPolicy setting due to
  https://github.com/PowerShell/PowerShell/issues/2742
- Add Known Issue that the HOME directory environment variable must be set
  explicitly due to:
  https://github.com/PowerShell/PowerShell/issues/1794

Tasks
- [x] Update doco
- [x] Add known issue
- [x] Add acceptance tests
- [x] Update travis to install PowerShell
- [x] Test acceptance on Ubuntu 14.04
- [x] Test acceptance on Ubuntu 16.04
- [x] Test acceptance on CentOS 7
- [x] Adhoc run

### CODE REVIEW

**Comment by glennsarti** [comment]
This PR should be merged in preference to https://github.com/puppetlabs/puppetlabs-powershell/pull/152 due to PowerShell 6.0 missing the required assemblies etc. to implement the PowerShell Manager.

This PR uses the legacy style of invocation (using a temp file) on Non-Windows platforms to give at least experimental support for PowerShell.  It is expected once the required assemblies etc. exist for the PowerShell Manager, a further PR will be raised to implement this on Non-Windows platforms.

**Comment by glennsarti** [comment]
This has been tested on an adhoc run to show it does not break existing Windows platforms, and has been run locally using VMPooler Beaker nodes for Ubuntu 14,16 and CentOS7.

Once the PR is merged, a further PR will be raised to add these platforms to the test matrix.

**Comment by glennsarti** [comment]
Note - This PR is only explicitly supporting Ubuntu 14/16 and CentOS 7 as these are platforms supported by PowerShell and have a an official distribution in a package repo.  It is expected that support for MacOSX will be added later.

**Comment by glennsarti** [comment]
Snippets of Adhoc results


2012R2 - 2017.1.x
```
19:36:32   Info: Applying configuration version '1490754992'
19:36:32   Notice: /Stage[main]/Main/Exec[no fail test]/returns: executed successfully
19:36:36   Notice: Applied catalog in 3.84 seconds
19:36:36 
19:36:36 qah1h4zghzfro5u.delivery.puppetlabs.net (windows2012r2-64-1) executed in 12.49 seconds
19:36:36 Exited: 2
19:36:36       should not fail
19:36:36 ssh connection to windows2012r2-64-1 has been terminated
19:36:36 ssh connection to redhat7-64-1 has been terminated
19:36:36 Handing 'qah1h4zghzfro5u.delivery.puppetlabs.net' back to vmpooler for VM destruction
19:36:36 Handing 'mgnkbq2t65xr4fx.delivery.puppetlabs.net' back to vmpooler for VM destruction
19:36:36 Spent 0.03 seconds cleaning up
19:36:36 
19:36:36 Finished in 4 minutes 45.7 seconds (files took 9 minutes 39 seconds to load)
19:36:36 29 examples, 0 failures
19:36:36 
19:36:37 + RESULT=0
19:36:37 + exit 0
19:36:37 [WS-CLEANUP] Deleting project workspace...[WS-CLEANUP] done
19:36:37 Finished: SUCCESS
```

2008R2 - 2016.5.x
```
19:53:01       should not fail
19:53:01 ssh connection to windows2008r2-64-1 has been terminated
19:53:01 ssh connection to redhat7-64-1 has been terminated
19:53:01 Handing 'oflk8hnwz5jxikp.delivery.puppetlabs.net' back to vmpooler for VM destruction
19:53:01 Handing 'kcr4bsb3rtv4dn1.delivery.puppetlabs.net' back to vmpooler for VM destruction
19:53:01 Spent 0.04 seconds cleaning up
19:53:01 
19:53:01 Finished in 5 minutes 2 seconds (files took 10 minutes 27 seconds to load)
19:53:01 29 examples, 0 failures
19:53:01 
19:53:01 + RESULT=0
19:53:01 + exit 0
19:53:01 [WS-CLEANUP] Deleting project workspace...[WS-CLEANUP] done
19:53:01 Finished: SUCCESS
```

2012R2 - Smoke Tests
```
19:21:31   Notice: /Stage[main]/Main/Exec[no fail test]/returns: executed successfully
19:21:35   Notice: Applied catalog in 3.64 seconds
19:21:35 
19:21:35 kqm0qaqywr2ijk1.delivery.puppetlabs.net (windows2012r2-64-1) executed in 13.14 seconds
19:21:35 Exited: 2
19:21:35       should not fail
19:21:35 ssh connection to windows2012r2-64-1 has been terminated
19:21:35 ssh connection to redhat7-64-1 has been terminated
19:21:35 Handing 'kqm0qaqywr2ijk1.delivery.puppetlabs.net' back to vmpooler for VM destruction
19:21:35 Handing 'r57868iwtfrstji.delivery.puppetlabs.net' back to vmpooler for VM destruction
19:21:35 Spent 0.04 seconds cleaning up
19:21:35 
19:21:35 Finished in 5 minutes 16 seconds (files took 1 minute 41.2 seconds to load)
19:21:35 29 examples, 0 failures
19:21:35 
19:21:35 + RESULT=0
19:21:35 + exit 0
19:21:35 [WS-CLEANUP] Deleting project workspace...[WS-CLEANUP] done
19:21:35 Finished: SUCCESS
```

**Comment by glennsarti** [comment]
I will run the acceptance tests on Ubuntu and CentOS manually and post results

**Comment by glennsarti** [comment]
Ubuntu 16.04
```
w92v9jw23qadnmw.delivery.puppetlabs.net (ubuntu1604-64-1) executed in 3.58 seconds
Exited: 2
      should not fail
ssh connection to ubuntu1604-64-1 has been terminated
ssh connection to redhat7-64-1 has been terminated
Handing 'w92v9jw23qadnmw.delivery.puppetlabs.net' back to vmpooler for VM destruction
Handing 'vlhe3nsf01p8rb1.delivery.puppetlabs.net' back to vmpooler for VM destruction
Spent 0.03 seconds cleaning up

Finished in 1 minute 23.21 seconds (files took 8 minutes 7 seconds to load)
27 examples, 0 failures
```

File: lib/puppet/provider/exec/powershell.rb [hunk 1]
```diff
@@ -3,13 +3,17 @@
...
+  commands :powershell => 
     if File.exists?("#{ENV['SYSTEMROOT']}\\sysnative\\WindowsPowershell\\v1.0\\powershell.exe")
       "#{ENV['SYSTEMROOT']}\\sysnative\\WindowsPowershell\\v1.0\\powershell.exe"
     elsif File.exists?("#{ENV['SYSTEMROOT']}\\system32\\WindowsPowershell\\v1.0\\powershell.exe")
       "#{ENV['SYSTEMROOT']}\\system32\\WindowsPowershell\\v1.0\\powershell.exe"
+    elsif File.exists?('/usr/bin/powershell')
+      '/usr/bin/powershell'
+    elsif File.exists?('/usr/local/bin/powershell')
+      '/usr/local/bin/powershell'
+    elsif !Puppet::Util::Platform.windows?
```
Comment (Iristyle) [inline lib/puppet/provider/exec/powershell.rb]: Do you really expect `powershell` to be somewhere in PATH but not in a `bin` directory?

File: .travis.yml [hunk 2]
```diff
@@ -1,12 +1,19 @@
...
+dist: trusty
 language: ruby
 cache: bundler
 script: "bundle exec rake release_checks"
 #Inserting below due to the following issue: https://github.com/travis-ci/travis-ci/issues/3531#issuecomment-88311203
 before_install:
   - gem update bundler
+  # Powershell installation instructions from
+  # https://github.com/PowerShell/PowerShell/blob/master/docs/installation/linux.md#ubuntu-1404
+  - curl https://packages.microsoft.com/keys/microsoft.asc | sudo apt-key add -
```
Comment (Iristyle) [inline .travis.yml]: `sudo` everywhere... 🐼

File: lib/puppet/provider/exec/powershell.rb (hunk 1 above)
Comment (glennsarti) [inline lib/puppet/provider/exec/powershell.rb]: No idea.  But you need a catchall somewhere.

**Pull request merged**